crates/flipperzero/examples/threads.rs

```diff
@@ -0,0 +1,65 @@
+//! Demonstrates use of threads on the Flipper Zero.
+
+#![no_main]
+#![no_std]
+
+// Required for panic handler
+extern crate flipperzero_rt;
+
+// Required for allocator
+extern crate flipperzero_alloc;
+
+extern crate alloc;
+
+use alloc::borrow::ToOwned;
+use core::time::Duration;
+
+use flipperzero::{furi::thread, println};
+use flipperzero_rt::{entry, manifest};
+
+// Define the FAP Manifest for this application
+manifest!(name = "Threads example");
+
+// Define the entry function
+entry!(main);
+
+// Entry point
+fn main(_args: *mut u8) -> i32 {
+    println!("Main app started!");
+
+    let first = thread::spawn(|| {
+        println!("First thread started!");
+        thread::sleep(Duration::from_secs(5));
+        println!("First thread finished!");
+        0
+    });
+
+    thread::sleep(Duration::from_secs(1));
+
+    let second = thread::Builder::new()
+        .name("Flipper".to_owned())
+        .expect("name is valid")
+        .spawn(|| {
+            println!("Second thread started!");
+            thread::sleep(Duration::from_secs(2));
+            println!("Second thread finished!");
+            0
+        });
+
+    for (i, thread) in [&thread::current(), first.thread(), second.thread()]
```

Comment:
Currently, if `thread::current()` is included in this list, I reliably crash Flipper Zero with a Null Pointer Dereference. I suspect it is because I'm freeing the `*mut sys::FuriThread` returned by `sys::furi_thread_get_current()`, which implies I don't currently fully understand how `FuriThread` memory is meant to be managed.

Reply:
That's interesting. I'm surprised it doesn't crash first from the `furi_assert(thread->state == FuriThreadStateStopped);`.

I think you'll need a boolean to keep track of whether you own this thread or it's a foreign one.

Reply:
I'm now doing this by never calling `sys::furi_thread_free` from `Thread` directly, but only inside either `impl Drop for JoinHandle` or in the thread's `FuriThreadStateStopped` handler.